release/nightly_tests/dataset/read_tfrecords_benchmark.py

```diff
@@ -0,0 +1,108 @@
+import random
+import shutil
+import tempfile
+from typing import List, Tuple
+
+import ray
+from ray.data.dataset import Dataset
+
+from benchmark import Benchmark
+from read_images_benchmark import generate_images
+import pyarrow as pa
+import numpy as np
+
+
+def read_tfrecords(path: str) -> Dataset:
+    return ray.data.read_tfrecords(paths=path)
+
+
+def generate_tfrecords_from_images(
+    num_images: int, sizes: List[Tuple[int, int]], modes: List[str], formats: List[str]
+) -> str:
+    images_dir = generate_images(num_images, sizes, modes, formats)
+    try:
+        ds = ray.data.read_images(images_dir)
+
+        # Convert images from NumPy to bytes
+        def images_to_bytes(batch):
+            images_as_bytes = [image.tobytes() for image in batch]
+            return pa.table({"image": images_as_bytes})
```

Comment:
I don't have a lot of context here, but why the bytes representation rather than our tensor extension type?

Reply:
Does `ds.write_tfrecords()` not work as expected with our tensor extension type?

Reply:
The row format for TFRecord (`tf,train.Example`) only support list of ints/floats/bytes ([doc](https://www.tensorflow.org/api_docs/python/tf/train/Example)):

```
Dict[str,
     Union[List[bytes],
           List[int64],
           List[float]]]
```

Is there a way to get bytes representation natively from tensor extension type?

Reply:
Anyways, since this isn't already implemented as I had expected, writing the ndarrays as bytes for this benchmark seems fine!

Reply:
@clarkzinzow re:

> And we could have support for NumPy ndarrays when converting Python values to TFRecord feature types

I am wondering how should we make sure the schema not lost here? We need to convert NumPy ndarray to bytes, right? Also we nee to add additional information.

Reply:
Yep TFRecords don't have support for schema/type metadata, unfortunately, so we would need to either:
1. serialize that metadata within the TFRecords data payload,
2. store a metadata sidecar file that contains this metadata, out-of-band for the TFRecord file format,
3. enforce that schema is specified on-read.

(2) is what TensorFlow Datasets does: https://www.tensorflow.org/datasets/external_tfrecord

We'd have pros and cons to consider for each option, e.g. (1) would make the data written by us only readable by us, which isn't great.